### ISSUE TITLE
[BUG] Labels only work in software

### ISSUE THREAD

**Issue by AdamJD01** [metallicafan212/M212-Editor-Issue-Tracker#47]
The labels for different things in a level only display when using software.

**Comment by metallicafan212** [comment]
They also work in DX7, but that's something that KW added. They draw directly on the viewport, and it's the only thing in the entire engine that does that. It needs a HDC to do the drawing, but it's completely incompatible with how both DX9 and GL work internally; it likes to draw the strings halfway through drawing the screen, meaning if I do give a HDC, it'll get overwritten by the contents of the screen, as both DX9 and GL do off screen rendering.

Honestly: it's not that important. I may fix it up in the future to make it work on all of them, but it basically does the same exact job my selection context box does, just honestly worse as it's input overload.

Closing for now as I don't intend on fixing this.

**Comment by metallicafan212** [comment]
Reopening as HP2 JP uses a native font, which means I do need to support the system....

**Comment by metallicafan212** [comment]
Fixed and implemented!

I changed the URenderDevice interface to add the DrawString method into it. Then, I test if the render device handles the call to it, and if so it skips the slow HDC rendering path.

In the DirectX9 render device I do all the font loading, rendering, text metrics, etc. and it results in text that draws on the screen with no performance penalty. Only problem is that it now draws over everything, but that only shows up with debugging turned on.
